onnxruntime/contrib_ops/rocm/bert/tunable_op.h

```diff
@@ -3,70 +3,116 @@
 
 #pragma once
 
-#include <map>
+#include <hip/hip_runtime.h>
+#include <hip/hip_fp16.h>
+
+#include <functional>
+#include <limits>
 #include <memory>
 #include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
 #include <vector>
-#include <hip/hip_runtime.h>
+
+#include "core/common/common.h"
 #include "contrib_ops/rocm/bert/util.h"
 
 namespace onnxruntime {
 namespace contrib {
 namespace rocm {
 
 struct OpParams {
+  OpParams() : stream{} {}
   explicit OpParams(hipStream_t stream) : stream(stream) {}
-  virtual std::string signature() const = 0;
```

Comment:
Shall we keep this, as it is required by for all derived classes? The code works without it, but I feel it is better to have it so that it is explicit to the users that they need to implement this function, to avoid the situation that users forget to implement it, and later realized by the compiler errors that this function is missing.

Reply:
The code will not work. It will be compile error. I might add some `static_assert`s to make it developer ergonomic.

Reply:
```cpp
template <typename ParamT>
class Op {
  template <typename T, typename = void>
  struct HasSignatureMethod : std::false_type {};
  template <typename T>
  struct HasSignatureMethod<T, typename std::enable_if_t<std::is_member_function_pointer<decltype(&T::Signature)>::value>> : std::true_type {};

  static_assert(HasSignatureMethod<ParamT>::value, "class derived from OpParams must implement public method 'Signature'");
//...
};
```

but will result 
```console
/home/guangyunhan/onnxruntime-1/onnxruntime/contrib_ops/rocm/bert/tunable_op.h:46:3: error: static_assert failed due to requirement 'HasSignatureMethod<onnxruntime::contrib::rocm::FastGeluParams<__half>, void>::value' "class derived from OpParams Must Implement Signature"
  static_assert(HasSignatureMethod<ParamT>::value, "class derived from OpParams Must Implement Signature");
  ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
and is still cryptic with hip clang

whereas the original without the `static_assert`
```cpp
/home/guangyunhan/onnxruntime-1/onnxruntime/contrib_ops/rocm/bert/tunable_op.h:114:36: error: no member named 'Signature' in 'onnxruntime::contrib::rocm::FastGeluParams<__half>'
      if (kernel_map_.find(params->Signature()) == kernel_map_.end()) {
                           ~~~~~~  ^
/home/guangyunhan/onnxruntime-1/onnxruntime/python/tools/kernel_explorer/kernels/fast_gelu.cc:41:24: note: in instantiation of member function 'onnxruntime::contrib::rocm::TunableOp<onnxruntime::contrib::rocm::FastGeluParams<__half>>::operator()' requested here
    ORT_THROW_IF_ERROR(op_(&params_));
```

concept is a c++20 feature, `static_assert` will not make it more ergonomic, so lets stick with the original one for now.